// angularfire2/utils/firebase_list_factory.js
var firebase_list_observable_1 = require('./firebase_list_observable');
var Firebase = require('firebase');
function FirebaseListFactory(absoluteUrl, _a) {
    var preserveSnapshot = (_a === void 0 ? {} : _a).preserveSnapshot;
    var ref = new Firebase(absoluteUrl);
    return new firebase_list_observable_1.FirebaseListObservable(function (obs) {
        var arr = [];
        ref.on('child_added', function (child) {
            arr = onChildAdded(arr, child);
            obs.next(preserveSnapshot ? arr : arr.map(unwrapMapFn));
        });
        ref.on('child_removed', function (child) {
            arr = onChildRemoved(arr, child);
            obs.next(preserveSnapshot ? arr : arr.map(unwrapMapFn));
        });
        ref.on('child_changed', function (child, prevKey) {
            arr = onChildChanged(arr, child, prevKey);
            obs.next(preserveSnapshot ? arr : arr.map(unwrapMapFn));
        });
        return function () { return ref.off(); };
    }, ref);
}
exports.FirebaseListFactory = FirebaseListFactory;
function unwrapMapFn(snapshot) {
    return snapshot.val();
}
function onChildAdded(arr, child) {
    var newArray = arr.slice();
    newArray.push(child);
    return newArray;
}
exports.onChildAdded = onChildAdded;
function onChildChanged(arr, child, prevKey) {
    return arr.reduce(function (accumulator, val, i) {
        if (!prevKey && i == 0) {
            accumulator.push(child);
            accumulator.push(val);
        }
        else if (val.key() === prevKey) {
            accumulator.push(val);
            accumulator.push(child);
        }
        else if (val.key() !== child.key()) {
            accumulator.push(val);
        }
        return accumulator;
    }, []);
}
exports.onChildChanged = onChildChanged;
function onChildRemoved(arr, child) {
    return arr.filter(function (c) { return c.key() !== child.key(); });
}
exports.onChildRemoved = onChildRemoved;
function onChildUpdated(arr, child, prevKey) {
    return arr.map(function (v, i, arr) {
        if (!prevKey && !i) {
            return child;
        }
        else if (i > 0 && arr[i - 1].key() === prevKey) {
            return child;
        }
        else {
            return v;
        }
    });
}
exports.onChildUpdated = onChildUpdated;
